lib/new_relic/telemetry_sdk/clients/client.rb

```diff
@@ -0,0 +1,76 @@
+# encoding: utf-8
+# This file is distributed under New Relic's license terms.
+# See https://github.com/newrelic/newrelic-telemetry-sdk-ruby/blob/main/LICENSE for complete details.
+# frozen_string_literal: true
+
+require 'net/http'
+require 'json'
+require 'zlib'
+require 'securerandom'
+
+module NewRelic
+  module TelemetrySdk
+    class Client
+      def initialize host:,
+                     path:,
+                     headers: {},
+                     # Note: see whether anything should be sent
+                     # via query params
+                     query_params: nil,
+                     use_gzip: true,
+                     payload_type:
+        @connection = set_up_connection host
+        @path = construct_full_path path, query_params
+        @headers = headers
+        @gzip_request = use_gzip
+        add_content_encoding_header @headers if @gzip_request
```

Comment:
I would prefer we make use of guard clauses within the `add_content_encoding_header` since we're passing `@header` in here vs. trailing `if` clauses.  From a maintainability point of view, the guard clauses are always checked without fail, but we may forget a trailing `if` when calling the methods.

Reply:
I don't have super strong feelings about this, but I feel like your point might apply more in the case of methods that we know will be used in multiple contexts - this one is very specific, and for that reason I don't think there is much risk in using an inline `if`.

Definitely not a pattern we should use blindly though - I see your point there.

Reply:
The way I think about it:
  * trailing conditionals should be use sparingly 
  * they're acceptable as a rule as part of guard clauses at the top of methods
  * anywhere else, ask the question: If the trailing condition is forgotten, will the called method still behave correctly?  If not, remove the trailing conditional and add a guard clause.

Reply:
Okay, I'll make this change. Thanks for the discussion about it.